### ISSUE TITLE
dart run default behavior is not consistent in Windows vs Linux

### ISSUE THREAD

**Issue by mythz** [dart-lang/sdk#44543]
I have a simple dart console project with these 2 files:

  bin/my_app.dart
  pubspec.yaml

#### bin\my_app.dart
```dart
void main(List<String> arguments) {
  print('Hello world!');
}
```

#### pubspec.yaml
```yml
name: my_app
description: A simple command-line application.

environment:
  sdk: '>=2.8.1 <3.0.0'
```

My folder name is a SHA hash, I'm able to run this dart project in Windows with:

    $ dart run

Because `my_app.dart` matches my package `my_app` name. 

But it fails if I try to run the same App in Linux:

```
Could not find the implicit file to run: bin/8a4f8b6ef9a0b2a6921e764840d49fc1b63e4698.dart.

Usage: dart run [arguments] <dart file | package target>
-h, --help                                             Print this usage information.
...
```

Where on Linux it requires the bin/file name to match the parent folder name and doesn't use the `bin/my_app.dart` file that matches the package name like windows.

Is there another way I can specify in my pubspec.yaml what is the default script I want `dart run` to run (e.g. "main" in node's package.json)?

**Comment by lrhn** [comment]
Just to be sure: Is the file name actually all lower-case? If not, it would explain why it can't be found on Linux, but can on Windows where the file system is case insensitive.

**Comment by mythz** [comment]
> Just to be sure: Is the file name actually all lower-case?

Yes, running this in an empty folder on Linux:

```sh
mkdir bin
cat <<EOF > bin/my_app.dart
void main(List<String> arguments) {
  print('Hello world!');
}
EOF

cat <<EOF > pubspec.yaml
name: my_app
description: A simple command-line application.

environment:
  sdk: '>=2.8.1 <3.0.0'
EOF

dart run
```

Results in:

```
Could not find the implicit file to run: bin/DartTest.dart.

Usage: dart run [arguments] <dart file | package target>
-h, --help                                             Print this usage information.

Debugging options:
    --observe=<[<port>[/<bind-address>]]>              The observe flag is a convenience flag used to run a program with a set of common options useful for debugging.

Options implied by --observe are currently:
    --enable-vm-service=<[<port>[/<bind-address>]]>    Enables the VM service and listens on the specified port for connections (default port number is 8181, default bind address is localhost).
    --[no-]pause-isolates-on-exit                      Pause isolates on exit when running with --enable-vm-service.
    --[no-]pause-isolates-on-unhandled-exceptions      Pause isolates when an unhandled exception is encountered when running with --enable-vm-service.
    --[no-]warn-on-pause-with-no-debugger              Print a warning when an isolate pauses with no attached debugger when running with --enable-vm-service.

Other debugging options:
    --[no-]pause-isolates-on-start                     Pause isolates on start when running with --enable-vm-service.
    --[no-]enable-asserts                              Enable assert statements.

Run "dart help" to see global options.
```

The same exact files in Windows results in:

```
Resolving dependencies...
Got dependencies!
Hello world!
```

**Comment by bkonyi** [comment]
This looks like there's an issue with the default script selection logic. Can you take a look or reassign to someone who can @jonasfj?

**Comment by jonasfj** [comment]
> Where on Linux it requires the bin/file name to match the parent folder name and doesn't use the bin/my_app.dart file that matches the package name like windows.

@mythz, this sounds like behavior we had in 2.10, which I think we fixed in 2.12 (currently in beta).

Please take a look, it should now be `dart run <package>:<script>` where `bin/<script>.dart` exists inside `<package>`.
And `dart run <package>` is a shorthand for `dart run <package>:<package>` which means run `bin/<package>.dart` from `package:<package>`.

Please reopen if this is still a concern in 2.12
